Mockito fails with JVM internal error when mocking class clinit does not find a transitive dependency

When tying to mock a class, where some transitive dependencies from the `<clinit>` are missing, Mockito swallows the root cause `NoClassDefFoundError`, but instead is reporting a JVM internal error
 `java.lang.InternalError: class redefinition failed: invalid class`, because the failing class is still tried to `retransform`.

Mockito reports the following exception:
```
org.mockito.exceptions.base.MockitoException: 
Mockito cannot mock this class: class org.mockitousage.bugs.creation.MockClassWithMissingStaticDepTest$ClassWithErrorInCinit.
Most likely it is a private class that is not visible by Mockito

You are seeing this disclaimer because Mockito is configured to create inlined mocks.
You can learn about inline mocks and their limitations under item #39 of the Mockito class javadoc.

	at org.mockito.internal.creation.bytebuddy.InlineDelegateByteBuddyMockMaker.prettifyFailure(InlineDelegateByteBuddyMockMaker.java:451)
	at org.mockito.internal.creation.bytebuddy.InlineDelegateByteBuddyMockMaker.createMockType(InlineDelegateByteBuddyMockMaker.java:416)
	at org.mockito.internal.creation.bytebuddy.InlineDelegateByteBuddyMockMaker.doCreateMock(InlineDelegateByteBuddyMockMaker.java:367)
	at org.mockito.internal.creation.bytebuddy.InlineDelegateByteBuddyMockMaker.createMock(InlineDelegateByteBuddyMockMaker.java:346)
	at org.mockito.internal.creation.bytebuddy.InlineByteBuddyMockMaker.createMock(InlineByteBuddyMockMaker.java:56)
	at org.mockito.internal.util.MockUtil.createMock(MockUtil.java:99)
	at org.mockito.internal.MockitoCore.mock(MockitoCore.java:84)
	at org.mockito.Mockito.mock(Mockito.java:2162)
	at org.mockito.Mockito.mock(Mockito.java:2077)
	at org.mockitousage.bugs.creation.MockClassWithMissingStaticDepTest.shouldNotSwallowCinitErrorsOfClassToMock(MockClassWithMissingStaticDepTest.java:21)
Caused by: java.lang.IllegalArgumentException: Could not create type
	at net.bytebuddy.TypeCache.findOrInsert(TypeCache.java:170)
	at net.bytebuddy.TypeCache$WithInlineExpunction.findOrInsert(TypeCache.java:399)
	at net.bytebuddy.TypeCache.findOrInsert(TypeCache.java:190)
	at net.bytebuddy.TypeCache$WithInlineExpunction.findOrInsert(TypeCache.java:410)
	at org.mockito.internal.creation.bytebuddy.TypeCachingBytecodeGenerator.mockClass(TypeCachingBytecodeGenerator.java:75)
	at org.mockito.internal.creation.bytebuddy.InlineDelegateByteBuddyMockMaker.createMockType(InlineDelegateByteBuddyMockMaker.java:408)
	... 47 more
Caused by: java.lang.InternalError: class redefinition failed: invalid class
	at java.instrument/sun.instrument.InstrumentationImpl.retransformClasses0(Native Method)
	at java.instrument/sun.instrument.InstrumentationImpl.retransformClasses(InstrumentationImpl.java:225)
	at org.mockito.internal.creation.bytebuddy.InlineBytecodeGenerator.triggerRetransformation(InlineBytecodeGenerator.java:285)
	at org.mockito.internal.creation.bytebuddy.InlineBytecodeGenerator.mockClass(InlineBytecodeGenerator.java:218)
	at org.mockito.internal.creation.bytebuddy.TypeCachingBytecodeGenerator.lambda$mockClass$0(TypeCachingBytecodeGenerator.java:78)
	at net.bytebuddy.TypeCache.findOrInsert(TypeCache.java:168)
	... 52 more
```


Test to reporduce the issue:

```java
public class MockClassWithMissingStaticDepTest {

    @Test
    public void shouldNotSwallowCinitErrorsOfClassToMock() {
        try {
            @SuppressWarnings("unused")
            var unused = Mockito.mock(ClassWithErrorInClinit.class);

            fail();
        } catch (MockitoException e) {
            e.printStackTrace();
            var cause = e.getCause();
            assertThat(cause).isInstanceOf(MockitoException.class);
            assertThat(cause.getMessage()).isEqualTo("Cannot instrument class org.mockitousage.bugs.creation.MockClassWithMissingStaticDepTest$ClassWithErrorInCinit because it or one of its supertypes could not be initialized");

            var cause2 = cause.getCause();
            assertThat(cause2).isInstanceOf(NoClassDefFoundError.class);
            assertThat(cause2.getMessage()).isEqualTo("Simulate missing transitive dependency used in <clinit>.");
        }
    }


    private static class ClassWithErrorInClinit {
        static {
            //noinspection ConstantValue
            if (true) {
                throw new NoClassDefFoundError("Simulate missing transitive dependency used in <clinit>.");
            }
        }
    }
}
```

## Root-Cause
The method `org.mockito.internal.creation.bytebuddy.InlineBytecodeGenerator.assureInitialization()` only handles `ExceptionInInitializerError` and swallows everything else.

## Expected behavior

Mockito shall report the first `NoClassDefFoundError` instead of swallowing it.

Affected version: `5.14.2`
Affected JVM: `21`

I will open a PR to fix the issue.

--- a/mockito-core/src/main/java/org/mockito/internal/creation/bytebuddy/InlineBytecodeGenerator.java
+++ b/mockito-core/src/main/java/org/mockito/internal/creation/bytebuddy/InlineBytecodeGenerator.java
@@ -234,12 +234,16 @@ public class InlineBytecodeGenerator implements BytecodeGenerator, ClassFileTran
     private static void assureInitialization(Class<?> type) {
         try {
             Class.forName(type.getName(), true, type.getClassLoader());
-        } catch (ExceptionInInitializerError e) {
+        } catch (Error error) {
+            Throwable ex = error;
+            if (ex instanceof ExceptionInInitializerError) {
+                ex = ((ExceptionInInitializerError) ex).getException();
+            }
             throw new MockitoException(
                     "Cannot instrument "
                             + type
                             + " because it or one of its supertypes could not be initialized",
-                    e.getException());
+                    ex);
         } catch (Throwable ignored) {
         }
     }

--- a/mockito-core/src/test/java/org/mockitousage/bugs/creation/MockClassWithMissingStaticDepTest.java
+++ b/mockito-core/src/test/java/org/mockitousage/bugs/creation/MockClassWithMissingStaticDepTest.java
@@ -1,0 +1,85 @@
+/*
+ * Copyright (c) 2024 Mockito contributors
+ * This program is made available under the terms of the MIT License.
+ */
+package org.mockitousage.bugs.creation;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.fail;
+
+import org.junit.Test;
+import org.mockito.Mockito;
+import org.mockito.exceptions.base.MockitoException;
+
+// See Issue #3498
+public class MockClassWithMissingStaticDepTest {
+
+    @Test
+    public void shouldNotSwallowClinitErrorOfClassToMock() {
+        try {
+            @SuppressWarnings("unused")
+            var unused = Mockito.mock(ClassWithErrorInClassInit.class);
+
+            fail();
+        } catch (MockitoException e) {
+            var cause = e.getCause();
+            assertThat(cause).isInstanceOf(MockitoException.class);
+            assertThat(cause.getMessage())
+                    .isEqualTo(
+                            "Cannot instrument class org.mockitousage.bugs.creation.MockClassWithMissingStaticDepTest$ClassWithErrorInClassInit because it or one of its supertypes could not be initialized");
+
+            var cause2 = cause.getCause();
+            assertThat(cause2).isInstanceOf(NoClassDefFoundError.class);
+            assertThat(cause2.getMessage())
+                    .isEqualTo("Simulate missing transitive dependency used in class init.");
+        } catch (NoClassDefFoundError ex) {
+            // Note: mock-maker-subclass will throw the NoClassDefFoundError as is
+            assertThat(ex.getMessage())
+                    .isEqualTo("Simulate missing transitive dependency used in class init.");
+        }
+    }
+
+    private static class ClassWithErrorInClassInit {
+        static {
+            //noinspection ConstantValue
+            if (true) {
+                throw new NoClassDefFoundError(
+                        "Simulate missing transitive dependency used in class init.");
+            }
+        }
+    }
+
+    @Test
+    public void shouldNotSwallowClinitExceptionOfClassToMock() {
+        try {
+            @SuppressWarnings("unused")
+            var unused = Mockito.mock(ClassWithExceptionInClassInit.class);
+
+            fail();
+        } catch (MockitoException e) {
+            var cause = e.getCause();
+            assertThat(cause).isInstanceOf(MockitoException.class);
+            assertThat(cause.getMessage())
+                    .isEqualTo(
+                            "Cannot instrument class org.mockitousage.bugs.creation.MockClassWithMissingStaticDepTest$ClassWithExceptionInClassInit because it or one of its supertypes could not be initialized");
+
+            var cause2 = cause.getCause();
+            assertThat(cause2).isInstanceOf(IllegalStateException.class);
+            assertThat(cause2.getMessage()).isEqualTo("Exception in class init.");
+        } catch (ExceptionInInitializerError ex) {
+            // Note: mock-maker-subclass will throw the ExceptionInInitializerError as is
+            var cause = ex.getCause();
+            assertThat(cause).isInstanceOf(IllegalStateException.class);
+            assertThat(cause.getMessage()).isEqualTo("Exception in class init.");
+        }
+    }
+
+    private static class ClassWithExceptionInClassInit {
+        static {
+            //noinspection ConstantValue
+            if (true) {
+                throw new IllegalStateException("Exception in class init.");
+            }
+        }
+    }
+}

